File: src/main/java/com/raychenon/hackerrank/warmup/BonAppetit.java
package com.raychenon.hackerrank.warmup;

import java.util.List;
import java.util.stream.IntStream;

/**
 * User: raychenon
 * Date: 2019-05-11
 * https://www.hackerrank.com/challenges/bon-appetit/problem
 */
public class BonAppetit {

    String calculateBillCharged(List<Integer> bill, int indexNotBilled, int amountCollected) {

        int amount = 0;
        int i = 0;
        amount = IntStream.range(0, bill.size())
                .filter(e -> bill.get(e) != bill.get(indexNotBilled))
                .map(x -> bill.get(x))
                .sum();

//        while (i < bill.size()) {
//            if (i != indexNotBilled) {
//                amount = amount + bill.get(i);
//            }
//            i++;
//        }

        amount = amount / 2;
        int amountOverCharged = amountCollected - amount;
        if (amountOverCharged == 0) {
            return "Bon Appetit";
        } else {
            return String.valueOf(amountOverCharged);
        }
    }

}

File: src/main/kotlin/com/raychenon/kotlin/hackerrank/warmup/BonAppetit.kt
package com.raychenon.kotlin.hackerrank.warmup

/**
 * User: raychenon
 * Date: 2019-05-10
 * https://www.hackerrank.com/challenges/bon-appetit/problem
 */
object BonAppetit {

    fun bonAppetit(bill: Array<Int>, k: Int, b: Int): Unit {
        var result = calculateBillCharged(bill, k, b)
        print("$result")
    }

    fun calculateBillCharged(bill: Array<Int>, indexNotBilled: Int, amountCharged: Int): String {
        var amount = 0
        for ((index, foodCost) in bill.withIndex()) {
            if (index != indexNotBilled) {
                amount = amount + foodCost
            }
        }

        amount = amount / 2
        if (amount == amountCharged) {
            return "Bon Appetit"
        } else {
            val overcharged = amountCharged - amount
            return "$overcharged"
        }
    }
}